pkg/ebpf/bytecode/runtime/asset.go

```diff
@@ -69,39 +69,114 @@ func (a *asset) Compile(config *ebpf.Config, additionalFlags []string, client st
+		return nil, fmt.Errorf("failed to create ram backed file from %s: %w", f.Name(), err)
+	}
+	defer closeFn()
+
+	if err = a.verify(inputFile); err != nil {
 		a.tm.compilationResult = verificationError
 		return nil, fmt.Errorf("error reading input file: %s", err)
 	}
 
-	out, result, err := compileToObjectFile(inputReader, outputDir, a.filename, a.hash, additionalFlags, kernelHeaders)
+	out, result, err := compileToObjectFile(inputFile, outputDir, a.filename, a.hash, additionalFlags, kernelHeaders)
 	a.tm.compilationResult = result
 
 	return out, err
 }
 
+// replace the symlink file with a copy of the input file so that
+// debug info can be resolved by tools like objdump
+func setupSourceInfoFile(source io.Reader, path string) error {
+	f, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	if _, err := io.Copy(f, source); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// creates a ram backed file from the given reader. The file is made immutable
+func createRamBackedFile(name, hash string, source io.Reader, runtimeDir string) (string, func(), error) {
+	var err error
+
+	if err = os.MkdirAll(runtimeDir, 0755); err != nil {
+		return "", nil, fmt.Errorf("unable to create compiler output directory %s: %w", runtimeDir, err)
+	}
+
+	memfdFile, err := memfd.CreateNameFlags(name, memfd.AllowSealing|memfd.Cloexec)
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to create memfd file: %w: ", err)
+	}
+	defer func() {
+		if err != nil {
+			memfdFile.Close()
+		}
+	}()
+
+	if _, err = io.Copy(memfdFile, source); err != nil {
+		return "", nil, fmt.Errorf("error copying bytes to memfd file: %w", err)
+	}
+
+	// seal the memfd file, making it immutable.
+	if err = memfdFile.SetSeals(memfd.SealAll); err != nil {
+		return "", nil, fmt.Errorf("failed to seal memfd file: %w", err)
+	}
+
+	target := fmt.Sprintf("/proc/%d/fd/%d", os.Getpid(), memfdFile.Fd())
+	tmpFile := filepath.Join(runtimeDir, fmt.Sprintf("%s-%s", name, hash))
+
+	link, _ := os.Readlink(target)
+	log.Debugf("createRamBackedFile: created memfd file: %s => %s", target, link)
+
+	os.Remove(tmpFile)
+	if err := os.Symlink(target, tmpFile); err != nil {
+		return "", nil, fmt.Errorf("failed to create symlink with target file %s: %w", target, err)
+	}
+
+	return tmpFile, func() {
+		os.Remove(tmpFile)
+
+		// io.Copy changes the file cursor. Reset it before use.
+		if _, err := memfdFile.Seek(0, os.SEEK_SET); err != nil {
+			log.Debug(err)
+		}
+		if err := setupSourceInfoFile(memfdFile, tmpFile); err != nil {
+			log.Debug("failed to setup source file: ", err)
+		}
+		memfdFile.Close()
+	}, nil
+}
+
 // verify reads the asset in the provided directory and verifies the content hash matches what is expected.
-// On success, it returns an io.Reader for the contents
-func (a *asset) verify(dir string) (io.Reader, error) {
-	p := filepath.Join(dir, "runtime", a.filename)
-	f, err := os.Open(p)
+func (a *asset) verify(verifyFile string) error {
```

Comment:
I think this is OK for the first version, but we should think about tweaking this so it can't be misused. The main reason it returned an `io.Reader` before was so you wouldn't accidentally introduce a TOCTOU vulnerability.

Since we are dealing with a sealed file now, this isn't a problem. Perhaps we should wrap a sealed file with an object, and only accept that here to prevent misuse. WDYT?

Reply:
Makes sense. The signature should convey the correct usage.
How does this work: https://github.com/DataDog/datadog-agent/pull/15726/commits/d57f5d8d3373044b9fd8a34841ac4638968d3f78?